hpx/lcos/dist_object.hpp

```diff
@@ -0,0 +1,566 @@
+        /// locality and meta object is essentailly a table that can find the
+        /// instances of dist_object in all localities. The All_to_All option only
+        /// locally holds the client and server of the dist_object.
+        /// \param base_name The name of the dist_object, which should be a unique
+        /// string across the localities
+        /// \param data The data of the type T of the dist_object
+        dist_object(std::string base, data_type data)
+          : base_type(create_server(data))
+          , base_(base)
+        {
+            HPX_ASSERT(C == construction_type::All_to_All ||
+                C == construction_type::Meta_Object);
+            if (C == construction_type::Meta_Object)
+            {
+                meta_object mo(base);
+                basename_list = mo.registration(this->get_id());
+                basename_registration_helper(base);
+            }
+            else
+            {
+                basename_registration_helper(base);
+            }
+        }
+
+        /// \cond NOINTERNAL
+        dist_object(hpx::future<hpx::id_type>&& id)
+          : base_type(std::move(id))
+        {
+        }
+
+        dist_object(hpx::id_type&& id)
+          : base_type(std::move(id))
+        {
+        }
+        /// \endcond
+
+        /// Access the calling locality's value instance for this dist_object
+        data_type const operator*() const
+        {
+            HPX_ASSERT(this->get_id());
+            ensure_ptr();
+            return **ptr;
+        }
+
+        /// Access the calling locality's value instance for this dist_object
+        data_type operator*()
+        {
+            HPX_ASSERT(this->get_id());
+            ensure_ptr();
+            return **ptr;
+        }
+
+        /// Access the calling locality's value instance for this dist_object
+        T const* operator->() const
+        {
+            HPX_ASSERT(this->get_id());
+            ensure_ptr();
+            return &**ptr;
+        }
+
+        /// Access the calling locality's value instance for this dist_object
+        T* operator->()
+        {
+            HPX_ASSERT(this->get_id());
+            ensure_ptr();
+            return &**ptr;
+        }
+
+        /// Asynchronously returns a future of a copy of the instance of this
+        /// dist_object associated with the given locality index. The locality
+        /// index must be a valid locality ID with this dist_object.
+        hpx::future<T> fetch(int idx)
+        {
+            HPX_ASSERT(this->get_id());
+            hpx::id_type lookup = get_basename_helper(idx);
+            typedef typename server::dist_object_part<T&>::fetch_ref_action
+                action_type;
+            return hpx::async<action_type>(lookup);
+        }
+        /// \cond NOINTERNAL
+    private:
+        mutable std::shared_ptr<server::dist_object_part<T&>> ptr;
+        std::string base_;
+        void ensure_ptr() const
```

Comment:
Why have this lazily? An instance to `dist_object` is always local (IIUC), so upon construction, we can have this pointer and save us the condition here.

Reply:
Good suggestion. I put the `ensure_ptr` inside the constructor now.
https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/hpx/lcos/distributed_object.hpp#L310
Let me know if you have any suggestion.

Reply:
Yes, and with #3827 merged this does not incur any undue overhead.